File: graph/Dijkstra/python/Dijkstra.py
from __future__ import annotations

from collections import deque
import heapq
from collections.abc import Sequence

MAX_DISTANCE = (1 << 63) - 1


def Dijkstra(graph: Sequence[Sequence[tuple[int, int]]], start: int) -> tuple[list[int], list[bool]]:
    graph_reachable = [False] * len(graph)
    traversal_queue: deque[int] = deque([start])
    graph_reachable[start] = True

    while traversal_queue:
        node = traversal_queue.popleft()
        for neighbor, _weight in graph[node]:
            if not graph_reachable[neighbor]:
                graph_reachable[neighbor] = True
                traversal_queue.append(neighbor)

    distances = [0] * len(graph)
    reachable = [False] * len(graph)
    min_heap: list[tuple[int, int]] = [(0, start)]
    reachable[start] = True
    distances[start] = 0

    while min_heap:
        distance, node = heapq.heappop(min_heap)

        if not reachable[node] or distance != distances[node]:
            continue

        for neighbor, weight in graph[node]:
            if distance > MAX_DISTANCE - weight:
                continue

            candidate = distance + weight
            if not reachable[neighbor] or candidate < distances[neighbor]:
                reachable[neighbor] = True
                distances[neighbor] = candidate
                heapq.heappush(min_heap, (candidate, neighbor))

    if any(is_graph_reachable and not is_reachable for is_graph_reachable, is_reachable in zip(graph_reachable, reachable)):
        raise OverflowError("Shortest-path overflow: a path distance exceeded the signed 64-bit integer range.")

    return distances, reachable

File: graph/Dijkstra/rust/src/Dijkstra.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;
use std::collections::VecDeque;

pub const MAX_DISTANCE: u64 = i64::MAX as u64;

pub struct DijkstraResult {
    pub distances: Vec<u64>,
    pub reachable: Vec<bool>,
}

#[allow(non_snake_case)]
pub fn Dijkstra(graph: &[Vec<(usize, u64)>], start: usize) -> Result<DijkstraResult, &'static str> {
    let mut graph_reachable = vec![false; graph.len()];
    let mut traversal_queue = VecDeque::from([start]);
    graph_reachable[start] = true;

    while let Some(node) = traversal_queue.pop_front() {
        for &(neighbor, _weight) in &graph[node] {
            if !graph_reachable[neighbor] {
                graph_reachable[neighbor] = true;
                traversal_queue.push_back(neighbor);
            }
        }
    }

    let mut distances = vec![0; graph.len()];
    let mut reachable = vec![false; graph.len()];
    let mut min_heap = BinaryHeap::new();

    reachable[start] = true;
    distances[start] = 0;
    min_heap.push((Reverse(0_u64), start));

    while let Some((Reverse(distance), node)) = min_heap.pop() {
        if !reachable[node] || distance != distances[node] {
            continue;
        }

        for &(neighbor, weight) in &graph[node] {
            let candidate = match distance.checked_add(weight) {
                Some(value) if value <= MAX_DISTANCE => value,
                _ => continue,
            };

            if !reachable[neighbor] || candidate < distances[neighbor] {
                reachable[neighbor] = true;
                distances[neighbor] = candidate;
                min_heap.push((Reverse(candidate), neighbor));
            }
        }
    }

    if graph_reachable
        .iter()
        .zip(reachable.iter())
        .any(|(&is_graph_reachable, &is_reachable)| is_graph_reachable && !is_reachable)
    {
        return Err("Shortest-path overflow: a path distance exceeded the signed 64-bit integer range.");
    }

    Ok(DijkstraResult {
        distances,
        reachable,
    })
}
